src/system/SystemPacketBuffer.h

```diff
@@ -653,7 +657,29 @@ inline PacketBufferHandle PacketBuffer::Last()
  *      // valid data
  *  @endcode
  */
-class PacketBufferWriter : public Encoding::LittleEndian::BufferWriter
+class PacketBufferWriterImpl
+{
+public:
+    /**
+     * Test whether this PacketBufferWriter is null, or conversely owns a PacketBuffer.
+     *
+     * @retval true     The PacketBufferWriter is null; it does not own a PacketBuffer. This implies either that
+     *                  construction failed, or that \c Finalize() has previously been called to release the buffer.
+     * @retval false    The PacketBufferWriter owns a PacketBuffer, which can be written using BufferWriter \c Put() methods,
+     *                  and (assuming no overflow) obtained by calling \c Finalize().
+     */
+    bool IsNull() const { return mPacket.IsNull(); }
+
+protected:
+    PacketBufferWriterImpl(size_t aAvailableSize, uint16_t aReservedSize, BufferWriter & aBufferWriter);
+    System::PacketBufferHandle Finalize(BufferWriter & aBufferWriter);
+
+private:
+    System::PacketBufferHandle mPacket;
+};
+
+template <class Writer>
+class PacketBufferWriterBase : public Writer, public PacketBufferWriterImpl
```

Comment:
Could we avoid this multiple implementation inheritance?

Reply:
Yes, but it'll still be ugly.